scope.go

```diff
@@ -0,0 +1,604 @@
+
+// Basic resource mamagement.
+type Resources struct {
+	limit Limit
+
+	nconnsIn, nconnsOut     int
+	nstreamsIn, nstreamsOut int
+	nfd                     int
+
+	memory  int64
+	buffers map[interface{}][]byte
+}
+
+// DAG ResourceScopes.
+// Resources accounts for the node usage, constraints signify
+// the dependencies that constrain resource usage.
+type ResourceScope struct {
+	sync.Mutex
+	done   bool
+	refCnt int
+
+	rc          *Resources
+	constraints []*ResourceScope
+}
+
+var _ network.ResourceScope = (*ResourceScope)(nil)
+var _ network.TransactionalScope = (*ResourceScope)(nil)
+
+func NewResources(limit Limit) *Resources {
+	return &Resources{
+		limit: limit,
+	}
+}
+
+func NewResourceScope(limit Limit, constraints []*ResourceScope) *ResourceScope {
+	return &ResourceScope{
+		rc:          NewResources(limit),
+		constraints: constraints,
+	}
+}
+
+// Resources implementation
+func (rc *Resources) checkMemory(rsvp int64) error {
+	// overflow check; this also has the side-effect that we cannot reserve negative memory.
+	newmem := rc.memory + int64(rsvp)
+	if newmem < rc.memory {
+		return fmt.Errorf("memory reservation overflow: %w", ErrResourceLimitExceeded)
+	}
+
+	// limit check
+	if newmem > rc.limit.GetMemoryLimit() {
+		return fmt.Errorf("cannot reserve memory: %w", ErrResourceLimitExceeded)
+	}
+
+	return nil
+}
+
+func (rc *Resources) releaseBuffers() {
+	for key, buf := range rc.buffers {
+		pool.Put(buf)
+		delete(rc.buffers, key)
+	}
+}
+
+func (rc *Resources) reserveMemory(size int64) error {
+	if err := rc.checkMemory(size); err != nil {
+		return err
+	}
+
+	rc.memory += int64(size)
+	return nil
+}
+
+func (rc *Resources) releaseMemory(size int64) {
+	rc.memory -= size
+
+	// sanity check for bugs upstream
+	if rc.memory < 0 {
+		panic("BUG: too much memory released")
+	}
+}
+
+func (rc *Resources) getBuffer(size int) ([]byte, error) {
+	if err := rc.checkMemory(int64(size)); err != nil {
+		return nil, err
+	}
+
+	buf := pool.Get(size)
+
+	rc.memory += int64(size)
+	if rc.buffers == nil {
+		rc.buffers = make(map[interface{}][]byte)
+	}
+	rc.buffers[buf] = buf
+
+	return buf, nil
+}
+
+func (rc *Resources) growBuffer(oldbuf []byte, newsize int) ([]byte, error) {
+	grow := newsize - len(oldbuf)
+	if err := rc.checkMemory(int64(grow)); err != nil {
+		return nil, err
+	}
+
+	_, ok := rc.buffers[oldbuf]
+	if !ok {
+		return nil, fmt.Errorf("cannot grow unknown buffer")
+	}
+
+	newbuf := pool.Get(newsize)
+	copy(newbuf, oldbuf)
+
+	rc.memory += int64(grow)
+	rc.buffers[newbuf] = newbuf
+	delete(rc.buffers, oldbuf)
+
+	return newbuf, nil
+}
+
+func (rc *Resources) releaseBuffer(buf []byte) {
+	_, ok := rc.buffers[buf]
+	if !ok {
+		panic("BUG: release unknown buffer")
+	}
+
+	rc.memory -= int64(len(buf))
+
+	// sanity check for bugs upstream
+	if rc.memory < 0 {
+		panic("BUG: too much memory released")
+	}
+
+	delete(rc.buffers, buf)
+	pool.Put(buf)
+}
+
+func (rc *Resources) addStream(incount, outcount int) error {
```

Comment:
iiuc, `incount` and `outcount` are always 0 and 1, or vice versa. Can we just pass the `network.Direction` here?

Reply:
addStream can be potentially used by some higher level migration, so I want to keep it as generic as possible.

Reply:
renamed this method to addStreams and added the simple addStream(dir), it indeed simplifies code quite a bit.